File: src/Cyborg/Utilities/Circle.cs
using System;
using System.Collections.Generic;
using Microsoft.Xna.Framework;
using static Cyborg.Utilities.MathsExtensions;

namespace Cyborg.Utilities
{
    public struct Circle
    {
        public Circle(Point centre, int radius)
        {
            Centre = centre;
            Radius = radius;
        }

        public Point Centre { get; }
        public int Radius { get; }

        public bool Intersects(Rectangle rectangle)
        {
            var centreDistanceX = Math.Abs(Centre.X - rectangle.Center.X);
            var centreDistanceY = Math.Abs(Centre.Y - rectangle.Center.Y);

            var isDefinitelyOutside = centreDistanceX > (rectangle.Width / 2 + Radius) || centreDistanceY > (rectangle.Height / 2 + Radius);
            if (isDefinitelyOutside)
                return false;

            var isDefinitelyInside = centreDistanceX <= rectangle.Width / 2 || centreDistanceY <= rectangle.Height / 2;
            if (isDefinitelyInside)
                return true;

            var cornerDistanceSquared = Math.Pow(centreDistanceX - rectangle.Width / 2, 2) + Math.Pow(centreDistanceY - rectangle.Height / 2, 2);
            return cornerDistanceSquared <= Math.Pow(Radius, 2);
        }

        public IEnumerable<Vector2> ToVertices(double verticesPerRadiusSquared = 0.25d)
        {
            var centreVector = Centre.ToVector2();
            var vertexCount = Math.Pow(Radius, 2) * verticesPerRadiusSquared;
            var angleIncrement = FullCircleAngle / vertexCount;

            for (var angle = 0d; angle < FullCircleAngle; angle += angleIncrement)
                yield return centreVector + Radius * new Vector2((float)Math.Cos(angle), (float)Math.Sin(angle));
        }
    }
}

File: src/common/utilities/circle.ts
import { centre, Rectangle } from "./rectangle.js";
import { subtract, Vector } from "./vector.js";

interface Circle extends Vector {
	radius: number;
}

interface Sector extends Circle {
	minimumAngle: number;
	maximumAngle: number;
}

const circleRectangleIntersects = (circle: Circle, rectangle: Rectangle): boolean => {
	const rectangleCentre = centre(rectangle);
	const centreDistanceX = Math.abs(circle.x - rectangleCentre.x);
	const centreDistanceY = Math.abs(circle.y - rectangleCentre.y);

	const isDefinitelyOutside = centreDistanceX > rectangle.width / 2 + circle.radius || centreDistanceY > rectangle.height / 2 + circle.radius;
	if (isDefinitelyOutside) return false;

	const isDefinitelyInside = centreDistanceX <= rectangle.width / 2 || centreDistanceY <= rectangle.height / 2;
	if (isDefinitelyInside) return true;

	const cornerDistanceSquared = Math.pow(centreDistanceX - rectangle.width / 2, 2) + Math.pow(centreDistanceY - rectangle.height / 2, 2);
	return cornerDistanceSquared <= Math.pow(circle.radius, 2);
};

const sectorRectangeIntersects = (sector: Sector, rectangle: Rectangle): boolean => {
	const rectangleCentre = centre(rectangle);
	const centreDistanceX = Math.abs(sector.x - rectangleCentre.x);
	const centreDistanceY = Math.abs(sector.y - rectangleCentre.y);

	const isDefinitelyOutside = centreDistanceX > rectangle.width / 2 + sector.radius || centreDistanceY > rectangle.height / 2 + sector.radius;
	if (isDefinitelyOutside) return false;

	// TODO this approach sucks
	const rectangleCorners = [
		{ x: rectangle.x, y: rectangle.y },
		{ x: rectangle.x + rectangle.width, y: rectangle.y },
		{ x: rectangle.x, y: rectangle.y + rectangle.height },
		{ x: rectangle.x + rectangle.width, y: rectangle.y + rectangle.height }
	];
	const rectangleCornersInsideAngle = rectangleCorners
		.map((corner) => {
			const vector = subtract(corner, sector);
			return (Math.atan2(vector.y, vector.x) + 2 * Math.PI) % (2 * Math.PI);
		})
		.filter((rad) =>
			sector.maximumAngle >= sector.minimumAngle
				? rad >= sector.minimumAngle && rad <= sector.maximumAngle
				: rad <= sector.minimumAngle || rad > sector.maximumAngle
		);

	return !!rectangleCornersInsideAngle.length;
};

export { Circle, Sector, circleRectangleIntersects, sectorRectangeIntersects };
